test/old_tests/UnitTests/Composable.cpp

```diff
@@ -167,4 +169,120 @@ TEST_CASE("Composable conversions")
+    {
+        struct stringable final : ABI::Windows::Foundation::IStringable
+        {
+            stringable(winrt::Windows::Foundation::IInspectable const& object, hstring const& value) :
+                m_object(object.as<::IInspectable>()),
+                m_value(value)
+            {
+            }
+
+            HRESULT __stdcall ToString(HSTRING* result) noexcept final
+            {
+                return WindowsDuplicateString(static_cast<HSTRING>(get_abi(m_value)), result);
+            }
+
+            HRESULT __stdcall QueryInterface(GUID const& id, void** result) noexcept final
+            {
+                if (is_guid_of<IStringable>(id))
+                {
+                    *result = static_cast<ABI::Windows::Foundation::IStringable*>(this);
+                    AddRef();
+                    return S_OK;
+                }
+
+                return m_object->QueryInterface(id, result);
+            }
+
+            ULONG __stdcall AddRef() noexcept final
+            {
+                return 1 + m_references.fetch_add(1, std::memory_order_relaxed);
+            }
+
+            ULONG __stdcall Release() noexcept final
+            {
+                uint32_t const remaining = m_references.fetch_sub(1, std::memory_order_relaxed) - 1;
+
+                if (remaining == 0)
+                {
+                    delete this;
+                }
+
+                return remaining;
+            }
+
+            HRESULT __stdcall GetIids(ULONG* count, GUID** iids) noexcept final
+            {
+                return m_object->GetIids(count, iids);
+            }
+
+            HRESULT __stdcall GetRuntimeClassName(HSTRING* result) noexcept final
+            {
+                return m_object->GetRuntimeClassName(result);
+            }
+
+            HRESULT __stdcall GetTrustLevel(::TrustLevel* result) noexcept final
+            {
+                return m_object->GetTrustLevel(result);
+            }
+
+        private:
+
+            com_ptr<::IInspectable> m_object;
+            hstring m_value;
+            std::atomic<uint32_t> m_references{ 1 };
+        };
+
+        *result = new (std::nothrow) stringable(object, value);
+        return *result ? S_OK : E_OUTOFMEMORY;
+    }
+}
+
+TEST_CASE("Composable tearoff")
+{
+    static std::array<winrt::guid, 1> tearoff_iids{ winrt::guid_of<IStringable>() };
+
+    struct Tearoff : DerivedT<Tearoff, IClosable>
+    {
+        void Close()
+        {
+        }
+
+        int32_t query_interface_tearoff(winrt::guid const& id, void** result) const noexcept final
+        {
+            if (is_guid_of<IStringable>(id))
+            {
+                return make_stringable(*this, L"ToString", result);
+            }
+
+            *result = nullptr;
+            return E_NOINTERFACE;
+        }
+
+        winrt::array_view<winrt::guid> get_iids_tearoff() const noexcept final
+        {
+            return winrt::array_view(tearoff_iids);
```

Comment:
What is the behavior when the author makes a mistake an `get_iids_tearoff` duplicates entries that are already on the list?

Reply:
The existing code already can create duplicate entries in the final array that is returned by `GetIids()`. So I'm guessing it isn't a big deal. [The docs](https://learn.microsoft.com/en-us/windows/win32/api/inspectable/nf-inspectable-iinspectable-getiids) say that it returns:

> an array that contains an IID for each interface implemented by the current Windows Runtime object. The [IUnknown](https://learn.microsoft.com/en-us/windows/desktop/api/unknwn/nn-unknwn-iunknown) and [IInspectable](https://learn.microsoft.com/en-us/windows/desktop/api/inspectable/nn-inspectable-iinspectable) interfaces are excluded.

So other than some minor inefficiencies of copying more than we need and having to check a few more items in the array, it seems pretty harmless.

Reply:
Seems reasonable to me.